src/transformers/models/gpt2/modeling_gpt2.py

```diff
@@ -487,6 +502,80 @@ def _set_gradient_checkpointing(self, module, value=False):
         if isinstance(module, GPT2Model):
             module.gradient_checkpointing = value
 
+    def prepare_inputs_for_generation(
```

Comment:
Let's not add this method to all GPT2 models - `GPT2ForTokenClassification` should not be able to call `generate()`

Reply:
Though I'm in favor of this option usually, I saw that all `PreTrainedModel` implement `GenerationMixin` which suggest that you could call `generate` with `GPT2ForTokenClassification`.

Would creating a specific generation class `GPT2GenerationMixin` that implements `GenerationMixin` be better? I don't know how this might conflict with GPT2 both implementing `PretrainedModel` and `GPT2GenerationMixin` (which both implement `GenerationMixin`).

Reply:
It might be a good idea to actually move `GenerationMixin` out of `PreTrainedModel` and to only add it to models that are actually be able to do generate. Happy to discuss this in a new issue.

Why should we though create a `GPT2GenerationMixin` ? This doesn't sound sensible to me. We can simple leave `prepare_inputs_for_generation` for both `GPT2DoubleHeadsModelOutput` and `GPT2LMHeadModel` as it was before - there is 0 problem with that. 

Note that we don't mind really if we have code duplication - that's one of the main philosophies of Transformers. We have re-implement the attention layer in 50+ models and this has worked well so far. 
It's far more important that we stay consistent with our APIs. `prepare_inptus_for_generation` is never implemented in the `...PretrainedModel` class, but always in the `...LMHeadModel` or `...ForCausalLM` class of each model type -> we should stick to this to not surprise a reader with a new code API.